### PR TITLE
Add command line apps to Python wheel

### PR DESCRIPTION
**Pull request by remia**
This PR adds command line applications (C++ and Python) to the Python wheels, some points to note:
* The wheel size is growing significantly with this change, I took some actions to reduce it like striping binaries and removing the test suite but it is still around [20MB per wheel](https://test.pypi.org/project/opencolorio/2.3.0.dev4/#files). We might gain some more by using `-Os` and/or LTO to compile the apps but the most efficient way would most likely be to compile OCIO dynamically and have Python package and apps link to it. I haven't tried that and expect it will bring complexity to the build process.
* The PR also contains some C++ warnings fix,
* Documentation build fix extracted to #1809 

Note that I have no prior experience adding scripts to Python packages, I'm open to any suggestion.

Wheels are available on the test PyPI, install with this command (at your own risks):
`pip install -i https://test.pypi.org/simple/ opencolorio==2.3.0.dev4`

### CODE REVIEW

File: include/OpenColorIO/OpenColorTypes.h
Comment (doug-walker) [inline include/OpenColorIO/OpenColorTypes.h:778]: Minor:  I suspect "VarsRoles" was a name for the group, but it doesn't make sense as a description of the code.  Same for the next few groups.  These sections already have an rst header, so it might be better to just remove all three doxygen group lines.

**Review by doug-walker — APPROVED** [review]

**Comment by remia** [comment]
Tagging @JeanChristopheMorinPerso in case you have time to check these changes, your feedback would be appreciated!

**Comment by JeanChristopheMorinPerso** [comment]
Thanks for the ping @remia! I'll take a look as soon as I have time. I'm guessing you'll want to merge this before the next release? If so, when is the next release planned (so that I can prioritize on my side).

File: src/bindings/python/package/command_line.py
Comment (JeanChristopheMorinPerso) [inline src/bindings/python/package/command_line.py:13]: `close_fds` makes me a little bit nervous.

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
This looks great @remia ! It's exactly what I would have done.

>The wheel size is growing significantly with this change, I took some actions to reduce it like striping binaries and removing the test suite but it is still around [20MB per wheel](https://test.pypi.org/project/opencolorio/2.3.0.dev4/#files). We might gain some more by using -Os and/or LTO to compile the apps but the most efficient way would most likely be to compile OCIO dynamically and have Python package and apps link to it. I haven't tried that and expect it will bring complexity to the build process.

What I did in the past was to compile dynamically, but linking the dependencies statically. So you end up with a single library that you can now link to the executables.

```mermaid
graph TD
    A[dep1.a] -->C(libocio.so)
    B[dep2.a] -->C(libocio.so)
    C --> D(app1)
    C --> E(app2)
    C --> F(OpenColorIO.cpython310.so python lib)
```
On Linux and macOS, it will work without any additional work. On WIndows, You'll need to use `os.add_dll_directory`.

File: src/bindings/python/package/command_line.py
Comment (remia) [reply src/bindings/python/package/command_line.py]: If memory serve, this was directly copied from [ninja python wheel](https://github.com/scikit-build/ninja-python-distributions/blob/master/src/ninja/__init__.py#L51), I'm not sure why they used this flag exactly so I removed in for now, thanks.

**Comment by remia** [comment]
Thanks for the review @JeanChristopheMorinPerso! I'll do some tests around the dynamic linking, I think I had some issues around the wheel repair steps in the past, will keep updating this PR to address the issues coming up.

**Comment by JeanChristopheMorinPerso** [comment]
@remia make sure that the dependencies are beside the OCIO shared library (which is why I suggested to statically link the dependencies to the OCIO shared library, since it's easier to bundle that way).

**Comment by remia** [comment]
The issue was just that on Windows the DLL for OCIO get into the bin folder on the wheel layout, so I updated the `__init__.py` to add this folder to the PATH. The OCIO build process installs all dependencies and build everything as static which is handy here.

I made a new test wheel on the [test PyPI](https://test.pypi.org/project/opencolorio/2.3.0.dev5/), I only checked it on mac OS but hopefully should work elsewhere if our tests are to be trusted.

**Comment by JeanChristopheMorinPerso** [comment]
Perfect. I'll manually inspect the wheels today on all platforms.

**Comment by JeanChristopheMorinPerso** [comment]
Perfect. I'll manually inspect the wheels today on all platforms.

**Comment by JeanChristopheMorinPerso** [comment]
On Linux, the wheels contain:

```
├── opencolorio.libs
│   └── libOpenColorIO-94504ed6.so
└── PyOpenColorIO
    ├── bin
    │   ├── ocioarchive
    │   ├── ociobakelut
    │   ├── ociocheck
    │   ├── ociochecklut
    │   ├── ocioconvert
    │   ├── ociolutimage
    │   ├── ociomakeclf
    │   ├── ocioperf
    │   ├── ociowrite
    │   ├── pyocioamf
    │   │   ├── config-aces-reference.yaml
    │   │   ├── example.amf
    │   │   ├── example_referenced_lut.clf
    │   │   ├── pyocioamf.py
    │   │   └── README.md
    │   └── pyociodisplay
    │       ├── pyociodisplay.py
    │       ├── README.md
    │       └── requirements.txt
    ├── command_line.py
    ├── __init__.py
    ├── libOpenColorIO.so
    └── PyOpenColorIO.so
```

`libOpenColorIO.so` seems to be duplicated (`opencolorio.libs/libOpenColorIO-94504ed6.so` which was created by repairwheel and `PyOpenColorIO/libOpenColorIO.so` which is the one generated by the actual build). When I run `diff <(nm --format just-symbols -C opencolorio.libs/libOpenColorIO-94504ed6.so) <(nm -C PyOpenColorIO/libOpenColorIO.so)`, both libraries have the same symbols.

The diff of `readelf -d` is
```
[jcmorin@arch01 asd]$ diff -u <(readelf -d opencolorio.libs/libOpenColorIO-94504ed6.so) <(readelf -d PyOpenColorIO/libOpenColorIO.so)
--- /dev/fd/63	2023-07-02 14:54:36.266401410 -0400
+++ /dev/fd/62	2023-07-02 14:54:36.269734762 -0400
@@ -1,22 +1,21 @@
 
-Dynamic section at offset 0x8d1608 contains 29 entries:
+Dynamic section at offset 0x6675c0 contains 28 entries:
   Tag        Type                         Name/Value
- 0x000000000000000f (RPATH)              Library rpath: [opencolorio.libs]
- 0x000000000000000e (SONAME)             Library soname: [libOpenColorIO-94504ed6.so]
  0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
  0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
  0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
  0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
+ 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN]
  0x000000000000000c (INIT)               0x83000
  0x000000000000000d (FINI)               0x51fab0
  0x0000000000000019 (INIT_ARRAY)         0x648438
  0x000000000000001b (INIT_ARRAYSZ)       992 (bytes)
  0x000000000000001a (FINI_ARRAY)         0x648818
  0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
- 0x000000006ffffef5 (GNU_HASH)           0x670028
- 0x0000000000000005 (STRTAB)             0x68c000
+ 0x000000006ffffef5 (GNU_HASH)           0x260
+ 0x0000000000000005 (STRTAB)             0xcd70
  0x0000000000000006 (SYMTAB)             0x2c60
- 0x000000000000000a (STRSZ)              99848 (bytes)
+ 0x000000000000000a (STRSZ)              99804 (bytes)
  0x000000000000000b (SYMENT)             24 (bytes)
  0x0000000000000003 (PLTGOT)             0x669000
  0x0000000000000002 (PLTRELSZ)           23952 (bytes)
```

The executables use `opencolorio.libs/libOpenColorIO-94504ed6.so`
```
 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN:$ORIGIN/../../opencolorio.libs]
 0x0000000000000001 (NEEDED)             Shared library: [libOpenColorIO-94504ed6.so]
```

And it looks like the osx wheels have the same problem.

On Windows, everything seems fine
```
C:.
└───PyOpenColorIO
    │   command_line.py
    │   PyOpenColorIO.pyd
    │   __init__.py
    │
    └───bin
        │   ocioarchive.exe
        │   ociobakelut.exe
        │   ociocheck.exe
        │   ociochecklut.exe
        │   ocioconvert.exe
        │   ociolutimage.exe
        │   ociomakeclf.exe
        │   ocioperf.exe
        │   ociowrite.exe
        │   OpenColorIO_2_3.dll
        │
        ├───pyocioamf
        │       config-aces-reference.yaml
        │       example.amf
        │       example_referenced_lut.clf
        │       pyocioamf.py
        │       README.md
        │
        └───pyociodisplay
                pyociodisplay.py
                README.md
                requirements.txt
```

I am able to import `PyOpenColorIO` without any issues.

**Comment by remia** [comment]
Thanks for checking @JeanChristopheMorinPerso! I had tested the wheel locally on mac OS but only using pip wheel, I will try to see why repairwheel (or mac OS equivalent) think it has to fix the wheel which seem like it was already working. Great catch though, I was wondering why Windows wheels were half the size.

**Review by cedrik-fuoco-adsk — APPROVED** [review]

**Comment by remia** [comment]
As a quick update, I think the reason why repairwheel / delocate is patching the generated wheel is because the disabling of the SONAME symlinks I added as a CMake build option when building the wheel is preventing CMake to generate the correct library install names (instead of @rpath/mylib.dylib I get a full path on mac OS for example). I did that because otherwise I get 3 copy of the OCIO lib in the wheel due to missing support for symlinks. On the flip side, I assume relying on a wheel repair process to get viable wheels is not really ideal, as it would presumably not work well for wheels build manually from source distribution.

I'll continue my investigations.

**Comment by remia** [comment]
I uploaded a new test release, `pip install -i https://test.pypi.org/simple/ opencolorio==2.3.0.dev6`, the wheels should not contain duplicated libs anymore.

File: setup.py
Comment (JeanChristopheMorinPerso) [inline setup.py:42]: I think if you remove the `SOVERSION` it shuld fix the problem. And anyway when building the wheel, it makes little sense to have the version embedded in the dynamic library file name.

File: setup.py
Comment (JeanChristopheMorinPerso) [inline setup.py:115]: Unrelated to this, but is there a reason why the docs are built here? It makes it so that to build a wheel, you need doxygen and sphinx installed.

File: setup.py
Comment (remia) [reply setup.py]: I made a flag to disable the symlink, this is much simpler indeed, thanks.

File: setup.py
Comment (remia) [reply setup.py]: This is used for the docstring, I made this optional (see also this issue #1744).

Comment (JeanChristopheMorinPerso) [reply setup.py]: I see. I'm not too sure if I understand what you mean. Anyway, my comment is unrelated to this PR. I just think that in general, building a wheel shouldn't involve documentation tools.

File: src/bindings/python/package/__init__.py
Comment (JeanChristopheMorinPerso) [inline src/bindings/python/package/__init__.py:33]: Unrelated to this PR (just a general comment): it would be better if this was off by default (i.e. opt-in, not opt-out).

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Great work @remia!

File: setup.py
Comment (JeanChristopheMorinPerso) [reply setup.py]: Indeed, it's now much cleaner!

File: setup.py
Comment (remia) [reply setup.py]: The documentation written in the C++ header will end up in the generated docstring for the Python wheel, developer can use help() to see the documentation, or the IDE can show it directly.

File: src/bindings/python/package/__init__.py
Comment (remia) [reply src/bindings/python/package/__init__.py:33]: I generally agree with that although I didn't investigate enough yet the pros and cons, I think we should decide the desired behaviour in the corresponding issue https://github.com/AcademySoftwareFoundation/OpenColorIO/issues/1785. When that's settle I can make the change, probably in a new PR to not block this one.

**Comment by remia** [comment]
Thanks @JeanChristopheMorinPerso, the PR is looking much nicer following your reviews!

**Comment by remia** [comment]
Last round of test wheels from the last update couple of days ago available: pip install -i https://test.pypi.org/simple/ opencolorio==2.3.0.dev7

**Comment by doug-walker** [comment]
@remia , would you resolve the conflict here and we will go ahead and merge this?  I notice that our Wheel GitHub Action has been failing somewhat randomly for the last two weeks, not sure if this will improve that?

**Comment by remia** [comment]
@doug-walker The latest fails I see on macOS are related to the recent SIMD updates:

```
    /Users/runner/work/OpenColorIO/OpenColorIO/src/OpenColorIO/CPUInfo.cpp:44:50: error: invalid output constraint '=a' in asm
        __asm__ volatile (".byte 0x0f, 0x01, 0xd0" : "=a"(eax), "=d"(edx) : "c" (index));
                                                     ^
    /Users/runner/work/OpenColorIO/OpenColorIO/src/OpenColorIO/CPUInfo.cpp:58:11: error: invalid output constraint '=a' in asm
            : "=a" (data[0]), "=S" (data[1]), "=c" (data[2]), "=d" (data[3])
```

It seems the earlier issues were solved by @cedrik-fuoco-adsk in #1826

**Pull request merged**
